File: src/benchling_api_utils.lua
-- Define a module table
local api_utils = {}

local mime = require("mime")
local http = require("ssl.https")
local ltn12 = require("ltn12")
local json = require("dkjson")

config_example = {
    domain = "tenant.benchling.com",
    api_key = "sk_XXXXXXXXXXXXXXXX",
    auth_type = "api_key", -- OR "oauth"
    project = "My Projects"
}

-- Function to encode credentials
local function encode_key(str)
    local key_encoded = "Basic " .. mime.b64(str)
    return key_encoded
end

-- URL-encode a string
local function url_encode(url)
    if type(url) ~= "string" then
        error("Unsupported type given")
    end
    local encoded_url = url:gsub("\n", "\r\n")
    encoded_url = encoded_url:gsub(
        "([^%w%-_%.%~])",
        function(c)
            return string.format("%%%02X", string.byte(c))
        end
    )
    
    return encoded_url
end

local function length(containable)
    local cnt

    if type(containable) == "string" then
        cnt = #containable
    elseif type(containable) == "table" then
        cnt = 0
        for _, _ in pairs(containable) do
            cnt = cnt + 1
        end
    else
        error("Unsupported type given")
    end

    return cnt
end

-- Merge two arrays into one
local function merge_nested_tables(a, b)
    local result = {}
    local index = 1

    -- Copy all elements from the first table 'a' to 'result'
    for i = 1, length(a) do
        result[index] = a[i]
        index = index + 1
    end

    -- Copy all elements from the second table 'b' to 'result'
    for i = 1, length(b) do
        result[index] = b[i]
        index = index + 1
    end

    return result
end

-- Converts kebab-case to camelCase
local function kebab_to_camel_case(str)
    local result = ""
    local capitalize_next = false

    for i = 1, length(str) do
        local char = str:sub(i, i)
        if char == "-" then
            capitalize_next = true
        else
            result = result .. (capitalize_next and char:upper() or char)
            capitalize_next = false
        end
    end

    return result
end

-- Send api request
local function api_request(url, credentials, method, payload, auth_type)
    local response = {}
    local auth_header

    if auth_type == "oauth" then
        auth_header = "Bearer " .. credentials
    else
        -- Default = API key
        auth_header = encode_key(credentials)
    end

    local success, status_code, headers, status_text = http.request{
        url = url,
        method = method,
        headers = {
            ["Accept"] = "application/json",
            ["Authorization"] = auth_header,
            ["Content-Type"] = payload and "application/json" or nil
        },
        body = payload,
        sink = ltn12.sink.table(response)
    }

    if not success then
        error("Request failed")
    end

    local response_content = json.decode(table.concat(response))
    return response_content
end

-- Handles paginated API requests with optional endpoint parameters
local function paginated_api_request(base_url, credentials, method, payload, endpoint_key, kwargs, auth_type)
    local response_body = {}
    local next_token = ""

    repeat
        local url = base_url
        local endpoint_params = {}

        if next_token ~= "" then
            table.insert(endpoint_params, "nextToken=" .. next_token)
        end

        if kwargs then
            for key, value in pairs(kwargs) do
                table.insert(endpoint_params, url_encode(key) .. "=" .. url_encode(value))
            end
        end

        if #endpoint_params > 0 then
            url = url .. "?" .. table.concat(endpoint_params, "&")
        end

        local response_content = api_request(url, credentials, method, payload, auth_type)
        
        if response_content[endpoint_key] then
            response_body = merge_nested_tables(response_body, response_content[endpoint_key])
        else
            response_body = response_content
        end

        next_token = response_content["nextToken"] or ""
    until next_token == ""

    return response_body
end

local function request_with_pagination(config, endpoint, method, payload, version, kwargs)
    version = version or "v2"

    -- Auto-detect key for paginated result extraction
    local endpoint_key_map = {
        ["v2"] = kebab_to_camel_case(endpoint),
        ["v3-alpha"] = "items"
    }

    local endpoint_key = endpoint_key_map[version]
    local credentials = config["api_key"]
    local auth_type = config["auth_type"] or "api_key"

    local base_url = string.format("https://%s/api/%s/%s", config["domain"], version, endpoint)

    local response_body = paginated_api_request(
        base_url,
        credentials,
        method,
        payload and json.encode(payload) or nil,
        endpoint_key,
        kwargs,
        auth_type
    )

    return response_body
end

local function query(config, endpoint, kwargs, version)
    return request_with_pagination(config, endpoint, "GET", nil, version, kwargs)
end

local function create(config, endpoint, payload, version)
    return request_with_pagination(config, endpoint, "POST", payload, version, nil)
end

local function update(config, endpoint, payload, version)
    return request_with_pagination(config, endpoint, "PATCH", payload, version, nil)
end

-- Get OAuth2 token using client_credentials
local function get_oauth_token(domain, client_id, client_secret)
    local token_url = string.format("https://%s/api/v2/token", domain)
    local credentials = client_id .. ":" .. client_secret
    local auth_header = "Basic " .. mime.b64(credentials)

    local response = {}
    local body = "grant_type=client_credentials"

    local success, status_code, headers, status = http.request{
        url = token_url,
        method = "POST",
        headers = {
            ["Authorization"] = auth_header,
            ["Content-Type"] = "application/x-www-form-urlencoded",
            ["Content-Length"] = tostring(#body),
        },
        source = ltn12.source.string(body),
        sink = ltn12.sink.table(response)
    }

    if not success or status_code ~= 200 then
        error("OAuth token request failed")
    end

    local decoded = json.decode(table.concat(response))
    return decoded["access_token"]
end
 
api_utils.query = query
api_utils.create = create
api_utils.update = update
api_utils.get_oauth_token = get_oauth_token

-- Export the module
return api_utils

File: src/benchling_api_utils.py
import base64
import json
import requests
from urllib.parse import urlencode
from typing import Any, Dict, Optional

def encode_key(api_key):
    return "Basic " + base64.b64encode(api_key.encode()).decode()

def kebab_to_camel_case(s):
    parts = s.split('-')
    return parts[0] + ''.join(word.capitalize() for word in parts[1:])

def api_request(url, credentials, method="GET", payload=None):
    headers = {
        "Accept": "application/json",
        "Authorization": encode_key(credentials),
    }
    if payload:
        headers["Content-Type"] = "application/json"

    response = requests.request(method, url, headers=headers, data=json.dumps(payload) if payload else None)
    response.raise_for_status()
    if not response.content:
        return {}
    return response.json()

def paginated_api_request(base_url, credentials, method, payload, endpoint_key, params=None):
    all_results = []
    next_token = None

    while True:
        query_params = params.copy() if params else {}
        if next_token:
            query_params["nextToken"] = next_token
        url = base_url + "?" + urlencode(query_params)
        data = api_request(url, credentials, method, payload)
        items = data.get(endpoint_key, [])
        all_results += items
        next_token = data.get("nextToken")
        if not next_token:
            break

    return all_results

def request_handler(config, endpoint, method="GET", payload=None, version="v2", params=None):
    domain = config["domain"]
    api_key = config["api_key"]
    base_url = f"https://{domain}/api/{version}/{endpoint}"
    if method == "GET":
        endpoint_key = "items" if version == "v3-alpha" else kebab_to_camel_case(endpoint)
        return paginated_api_request(base_url, api_key, method, payload, endpoint_key, params)

    if params:
        base_url += "?" + urlencode(params)
    return api_request(base_url, api_key, method, payload)

def query(config, endpoint, params=None, version="v2"):
    return request_handler(config, endpoint, "GET", None, version, params)

def create(config, endpoint, payload, version="v2"):
    return request_handler(config, endpoint, "POST", payload, version)

def update(config, endpoint, payload, version="v2"):
    return request_handler(config, endpoint, "PATCH", payload, version)

def post(config, endpoint, payload, version="v2"):
    return request_handler(config, endpoint, "POST", payload, version)

def get(config, endpoint, params=None, version="v2"):
    domain = config["domain"]
    api_key = config["api_key"]
    base_url = f"https://{domain}/api/{version}/{endpoint}"
    if params:
        base_url += "?" + urlencode(params)
    return api_request(base_url, api_key, "GET")

def get_oauth_token(domain, client_id, client_secret):
    url = f"https://{domain}/api/v2/token"
    credentials = f"{client_id}:{client_secret}"
    headers = {
        "Authorization": "Basic " + base64.b64encode(credentials.encode()).decode(),
        "Content-Type": "application/x-www-form-urlencoded"
    }
    body = {"grant_type": "client_credentials"}
    response = requests.post(url, headers=headers, data=body)
    response.raise_for_status()
    return response.json()["access_token"]

def get_project_id(config: Dict[str, Any], project_name: Optional[str] = None) -> str:
    target_name = project_name or config.get("project")
    projects = query(config, "projects")
    for project in projects:
        if project.get("name") == target_name:
            return project.get("id", "")
    return ""

def get_schema_id(config: Dict[str, Any], schema_name: str) -> str:
    schemas = query(config, "entity-schemas")
    for schema in schemas:
        if schema.get("name") == schema_name:
            return schema.get("id", "")
    return ""

__all__ = [
    "api_request",
    "create",
    "encode_key",
    "get",
    "get_oauth_token",
    "get_project_id",
    "get_schema_id",
    "kebab_to_camel_case",
    "paginated_api_request",
    "post",
    "query",
    "request_handler",
    "update",
]
